Sources/System/FileLock.swift

```diff
@@ -0,0 +1,74 @@
+#if !os(Windows)
+extension FileDescriptor {
+  /// Apply an advisory lock to the file associated with this descriptor.
+  ///
+  /// Advisory locks allow cooperating processes to perform consistent operations on files,
+  /// but do not guarantee consistency (i.e., processes may still access files without using advisory locks
+  /// possibly resulting in inconsistencies).
+  ///
+  /// The locking mechanism allows two types of locks: shared locks and exclusive locks.
+  /// At any time multiple shared locks may be applied to a file, but at no time are multiple exclusive, or
+  /// both shared and exclusive, locks allowed simultaneously on a file.
+  ///
+  /// A shared lock may be upgraded to an exclusive lock, and vice versa, simply by specifying the appropriate
+  /// lock type; this results in the previous lock being released and the new lock
+  /// applied (possibly after other processes have gained and released the lock).
```

Comment:
I appreciate this is as specific as `man 2 flock` gets and its implicit that the upgrade only works on the same file descriptor I suppose, i.e. if a process has two independent file descriptors to the same underlying file, they can still block each other.

Reply:
I believe `flock(2)` would be associated between process/file, while `OFD` locks would be file-description associated (but not descriptor, i.e. dup would share locks).

Reply:
We may already be on the same page here, but I believe that, with `flock(2)`, using `dup(2)` will allow the new fd to upgrade a lock, but using `open(2)` will not.

Here's a toy program that accepts `NEW_FD_METHOD` as an environment variable and attempts to upgrade a shared lock to an exclusive one using a new file descriptor pointing to the same file as was already locked:

https://gist.github.com/simonjbeaumont/7bd12f1f97e162734b6fec847d3a969f

Running it with both options and `dup` is fine, `open` fails:

```console
❯ clang -o flock-test flock-test.c && NEW_FD_METHOD=dup ./flock-test

❯ clang -o flock-test flock-test.c && NEW_FD_METHOD=open ./flock-test
error obtaining exclusive lock with new fd 3: Resource temporarily unavailable
```